include/OpenSpaceToolkit/Astrodynamics/Flight/Maneuver.hpp

```diff
@@ -0,0 +1,251 @@
+    );
+
+    /// @brief Output stream operator
+    ///
+    /// @code{.cpp}
+    ///                  std::cout << Maneuver(...) ;
+    /// @endcode
+    ///
+    /// @param anOutputStream An output stream
+    /// @param aManeuver A Maneuver
+    ///
+    /// @return A reference to output stream
+    friend std::ostream& operator<<(std::ostream& anOutputStream, const Maneuver& aManeuver);
+
+    /// @brief Check if the maneuver is defined
+    ///
+    /// @return True if maneuver is defined (always returns true)
+    bool isDefined() const;
+
+    /// @brief Get the instants of the maneuver
+    /// @return The instants
+    Array<Instant> getInstants() const;
+
+    /// @brief Get the acceleration profile of the maneuver
+    ///
+    /// @param aFrameSPtr A frame in which the acceleration profile is to be defined
+    ///
+    /// @return The acceleration profile
+    Array<Vector3d> getAccelerationProfile(const Shared<const Frame>& aFrameSPtr = DefaultAccelFrameSPtr) const;
+
+    /// @brief Get the mass flow rate profile of the maneuver
+    /// @return The mass flow rate profile
+    Array<Real> getMassFlowRateProfile() const;
+
+    /// @brief Get the start instant of the maneuver
+    /// @return The start instant
+    Instant getStartInstant() const;
+
+    /// @brief Get the end instant of the maneuver
+    /// @return The end instant
+    Instant getEndInstant() const;
+
+    /// @brief Get the duration of the maneuver
+    /// @return The duration
+    Duration getDuration() const;
+
+    /// @brief Calculate the deltaV (scalar metric) imparted during the maneuver
+    /// @return The deltaV
+    Scalar calculateDeltaV() const;
+
+    /// @brief Calculate the delta Mass lost during the maneuver
+    /// @return The delta Mass
+    Mass calculateDeltaMass() const;
+
+    /// @brief Calculate the average thrust imparted during the maneuver
+    ///
+    /// @param anInitialSpacecraftMass The initial mass of the spacecraft
+    ///
+    /// @return The average thrust
+    Scalar calculateAverageThrust(const Mass& anInitialSpacecraftMass) const;
+
+    /// @brief Calculate the average specific impulse produced during the maneuver
+    ///
+    /// @param anInitialSpacecraftMass The initial mass of the spacecraft
+    ///
+    /// @return The average specific impulse
+    Scalar calculateAverageSpecificImpulse(const Mass& anInitialSpacecraftMass) const;
+
+    /// @brief Calculate the (interpolated) acceleration at a given instant during the maneuver
+    ///
+    /// @param anInstant The instant at which to calculate the acceleration
+    /// @param aFrameSPtr The frame in which the acceleration is to be defined
+    /// @param anInterpolationType The interpolation type to use
+    ///
+    /// @return The acceleration
+    Vector3d calculateAccelerationAt(
+        const Instant& anInstant,
+        const Shared<const Frame>& aFrameSPtr,
+        const Interpolator::Type& anInterpolationType = DEFAULT_MANEUVER_INTERPOLATION_TYPE
```

Comment:
We shouldn't be live constructing the interpolator everytime this function is called. It's O(N) to construct a barycentric rational interpolator. And O(N) to call it.

Reply:
@vishwa2710 How else would you recommend it be done? 

The only way I could imagine how to avoid this is to have an interpolator type supplied as an argument as construction time of the `Maneuver` class and then stored as a member variable. Then you could leverage that interpolator (with its internally stored x's and y's) for each interpolation call. However you recommended against doing this in a previous comment.

I agree, if a user calls the `calculateAccelerationAt` function in a loop it will be slower, but that's what the `calculateAccelerationsAt` function is for (similar to this pattern in the `Propagator.calculateState(s)At`). The user is unlikely to call the `calculateAccelerationsAt` function multiple times in a loop, and the maneuver object itself will not be used in propagation (TabulatedDynamics will), so I don't think we should prematurely optimize for performance here.

Ditto for your comments on `calculateMassFlowRate(s)At` below (resolved those threads so we can keep the discussion on this one). Let me know what you think

Reply:
In general I would agree that we should not prematurely optimize. But from a user perspective there is no knowledge that one is more performant than the other or any such contract.
If we provide a way to calculate accelerations, users can use it do what they see fit.

1 way of dealing with this is to indeed provide an interpolation type that goes with the profile to specify how the profile should be interpolated, I'm not a huge fan of this.

Another way of handling this is to remove the `calculateAcceleration*` methods and just enforce users to create a tabulated dynamics and compute their accelerations from that.
In this way users are forced to reason about the interpolation type at the correct interface, and are guaranteed to get performance as the interpolator is created once and the right time.

We can also forgo the performance issues and leave it as is. IMO Maneuvers being mostly data objects, we are overloading the interpolation and "calculation" concepts whereas we should be separating our concerns. If you want to interpolate, create a tabulated dynamics. But feel free to seek further opinions from the team, or resolve it if you feel strongly.

Reply:
@vishwa2710 I see what you mean, that makes sense. I don't like the idea of  removing the calculateAcceleration* methods because they add a good amount of convenience for the user for plotting and stuff.

I think this "warning about not optimized performance" point is not that critical, because we already have something like this out [here](https://github.com/open-space-collective/open-space-toolkit-astrodynamics/blob/main/src/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.cpp#L139) for propagator.calculateStateAt where a user can call it over and over in a loop and it will reduce and expand the state between each call which not great efficiency wise.

I think the potential for misuse of performance decrease in this case (as in the propagator case) is outweighed by the convenience of having both the calculateXAt and calculateXsAt methods that we can leverage in the services, etc.

However, happy to have a team consensus of this sort of debate moving forward. @Derollez @phc1990 @kyle-cochran any opinions on this?

Also I'm curious why you're not a fan of the maneuver class having its own interpolator (from a software correctness interface perspective)

Reply:
Maneuver to me is a data object. In general, we have the ability move "up" all methods from child objects to the parent. However that is not great practice as if and when the interfaces change you are locked into multiple levels of breaking interfaces.
Ideally you just return a child object and allow the user to manipulate the information that they want.

In this case, I already feel like we are tacking on too much to the Maneuver object, if a user wants to interpolate in a specific way it's quite simple to:
`maneuver.to_tabulated_dynamics().compute_contribution(instant, Frame.GCRF())`

The added "simplicity" of a `calculateThrustAccelerationAt` IMO is outweighed by the mixing of a data model with a computational model, additional unit tests, boiler plate code, code maintenance and updates etc.

For example, if a class has an interval as a member variable, we don't additionally save the duration, we just return the interval, and then allow the user the get the duration from the interval as that interface is locked in and not the responsibility of the original class.

Reply:
Ok yes that makes a lot of sense, thanks for the clear explanation. I'm actually in favor of removing the `calculateAcceleration(s)At` and `calculateMassFlowRate(s)At` methods to force users to use `TabulatedDynamics` to interpolate.

But now, just to play devil's advocate here 🗡️ , there's no plural version  of `computeContribution`, so the user would have to loop over it to interpolate at multiple points which would be slower if the non-default (non-eci) frame is requested because a new frame would be created on every call. What do you think about this? I don't think its a big deal, but just something I noticed

Reply:
> For example, if a class has an interval as a member variable, we don't additionally save the duration, we just return the interval, and then allow the user the get the duration from the interval as that interface is locked in and not the responsibility of the original class.

This is a good point and great example to illustrate it.

> there's no plural version of computeContribution, so the user would have to loop over it to interpolate at multiple points which would be slower

That's also a good point.

I think you guys made some good points. I understand where Vishwa is coming from to keep objects and methods as elementary and modular as possible and give the responsibility to the user to introduce some performance hits. What about this to move forward on the short term -> We keep it simple and modular, remove those methods for now, deploy and start using in the service and in notebooks. We assess the potential pain or disconnect at user level and if it's worth it, we can always come back here and add the code (which already exists ^^).

@phc1990 Let us know if you have an opinion on this: You might have a different angle from the Guidance service development :)

Reply:
Actually Pau's comment below on different lines:
> If we leave implementation aside, just from a "design" perspective, if a maneuver is a historical set of time-based numbers, it should be just that.

Any byproduct of that (for instance, a tabulated dynamics) could be created with many different "flavours" (the interpolator type being one of them for now). Having these "biproducts" as memvars "forces" them to be frozen with one "flavour" or another, whereas the "beauty" (imo) of the Maneuver object is that you store only what you need to recreate said bi-products.


Seems to indicate that he is leaning towards the same conclusion but let's confirm and move 🙏 :)

Reply:
Just posting a new comment, since it's a slightly different but related topic. 

I think there is yet a 3rd group of functions that were considering:

- "Non-deterministic" getters/computer: this involves the computation of what might look like unambiguous properties, but, in reality, they are not. Examples of these are:
  - calculateDeltaV
  - calculateDeltaMass
  - calculateAverageThrust
  - calculateAverageSpecificImpulse
  - etc.

Whilst form the outside, one could do `maneuver.calculateDeltaV()` and simply get a number, I think this might be confusing:

On the one hand, we are providing a "raw but nice" way to the user to create a `TabulatedDynamics` where they are in full control with arguments such as the interpolator type...

...and on the other hand, we are "hiding" that we are using that behind each `calculateX` function an Euler type numerical integration is taking place.

What I would suggest:

1. We keep this "compute" functions to the bare minimum (for now) (e.g. are averageThrust and averageSpecificImpulse needed atm?, if so great 👍 )
2. For the functions we do know we actually need (maybe calculateDeltaV, calculateDeltaMass?), we already make the API explicit: we go from `calculateX()` to `calculateY(anIntegrationType)`. It does not matter if we only offer an EULER_TYPE integration at the moment, but this way we are consistent.
3. In the future (definitively not now), we should explore the option of moving the numerical integration capabilities to the `TabulatedDynamics` class. E.g:
  `tabulatedDynamcis.integrateContribution(aSubset , anIntegrationType)`

Reply:
(Apologies if these comments come in late, I lost track of this PR. All in all, the simpler we keep the `ManeuverObject` the better imo)

Reply:
Great points guys, thanks for the feedback. 
- Agreed, I will removed the `calculateX(s)At(instant(s))` methods
- I will change the getters to be accessors following our slack thread (unrelated to the conversation above and doesn't change the behavior at all)
- I see your point @phc1990 about being more explicit regarding the fact that we're actually doing euler integration inside the calculateX methods, but I feel like specifying that is overkill at this point. These methods will be used in the guidance service to compute metrics and analyze the maneuvers that are planned or have been performed, and I'm not sure we will need to make them more accurate by upping the order of the integration happening under the hood. That being said, I like your idea of adding a  `tabulatedDynamcis.integrateContribution(aSubset , anIntegrationType)` in a future refactor when we have more time, I think that's very clean. Also, these methods can be used by the mission analysis team for convenience as well. Therefore, I plan on keeping these four `calculate` methods for now cause I think they will be useful in the guidance service, but I'm happy to remove them in the future if they become confusing. Let me know if you agree with this.

Reply:
All good on my end @apaletta3 👍